internal/resources/cluster/nodepools/data_source_node_pool.go

```diff
@@ -13,21 +13,40 @@ import (
 	"github.com/pkg/errors"
 
 	"github.com/vmware/terraform-provider-tanzu-mission-control/internal/authctx"
+	clienterrors "github.com/vmware/terraform-provider-tanzu-mission-control/internal/client/errors"
+	"github.com/vmware/terraform-provider-tanzu-mission-control/internal/helper"
 	nodepoolsmodel "github.com/vmware/terraform-provider-tanzu-mission-control/internal/models/cluster/nodepool"
 	"github.com/vmware/terraform-provider-tanzu-mission-control/internal/resources/common"
 )
 
 func DataSourceClusterNodePool() *schema.Resource {
 	return &schema.Resource{
-		ReadContext: dataSourceClusterNodePoolRead,
-		Schema:      nodePoolSchema,
+		ReadContext: func(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+			return dataSourceClusterNodePoolRead(helper.GetContextWithCaller(ctx, helper.DataRead), d, m)
+		},
+		Schema: nodePoolSchema,
 	}
 }
 
-func dataSourceClusterNodePoolRead(_ context.Context, d *schema.ResourceData, m interface{}) (diags diag.Diagnostics) {
+func dataSourceClusterNodePoolRead(ctx context.Context, d *schema.ResourceData, m interface{}) (diags diag.Diagnostics) {
 	config := m.(authctx.TanzuContext)
 
-	resp, err := config.TMCConnection.NodePoolResourceService.ManageV1alpha1ClusterNodePoolResourceServiceGet(constructFullName(d))
+	fn := constructFullName(d)
+	resp, err := config.TMCConnection.NodePoolResourceService.ManageV1alpha1ClusterNodePoolResourceServiceGet(fn)
+
+	if err != nil {
+		if clienterrors.IsNotFoundError(err) {
+			if helper.IsDataRead(ctx) {
```

Comment:
here we need to flip the condition and the below return with error message will take care of this condition as well